### PR TITLE
RetroFuturaGradle port

### LINKED ISSUE #64: genIntellijRuns problems and a hacky fix
It seems like `genIntellijRuns` may cause more issues than help for the way this build script is set up?

In some tutorials it's suggested that you should run gradle task during IDE setup. It creates two application configurations like so:
image

The benefit is you can add more CLI arguments to the launch, however, when you run the tasks it skips over the token replacement section.

As a result if you use this method you end up with modinfo like: 
image

And end up with potential crashes like the following with the default proxy setup:
```
cpw.mods.fml.common.LoaderException: java.lang.ClassNotFoundException: GRADLETOKEN_GROUPNAME.ClientProxy
...
Caused by: java.lang.ClassNotFoundException: GRADLETOKEN_GROUPNAME.ClientProxy


cpw.mods.fml.common.LoaderException: java.lang.ClassNotFoundException: GRADLETOKEN_GROUPNAME.CommonProxy
...
Caused by: java.lang.ClassNotFoundException: GRADLETOKEN_GROUPNAME.CommonProxy
```

### Solutions to the task existing
I don't know what I'm doing with gradle build scripts honestly, if it's possible to make it so the application configurations also run the token replacement they're currently missing
```
    if (replaceGradleTokenInFile) {
        replaceIn replaceGradleTokenInFile
        if (gradleTokenModId) {
            replace gradleTokenModId, modId
        }
        if (gradleTokenModName) {
            replace gradleTokenModName, modName
        }
        if (gradleTokenVersion) {
            replace gradleTokenVersion, modVersion
        }
        if (gradleTokenGroupName) {
            replace gradleTokenGroupName, modGroup
        }
    }
```

It may be worth just throwing a deprecation warning any time you run `genIntellijRuns` pointing people to the other way.

### The other way
1. In Intellij on gradle menu click on Tasks → forgegradle → runClient (It doesn't matter if you want to debug)
image
2. Now click on the task dropdown and select `Edit Configurations`
image
3. And hit the little Save icon to prevent the task from running away.
image
4. You can now select that task and use Debug to hot swap code with Ctrl+Shift+F9 to your hearts content.
image

The only problem with this version is you can't add CLI arguments as easily, but if you really need to you can edit `build.gradle` and edit the following section, for example:
```gradle
runClient {
    if (developmentEnvironmentUserName) {
        arguments += [
            "--username", developmentEnvironmentUserName,
            "--uuid", "7657b3d4cfd34078b523df0e1eb0f307",
            "--width", "1280",
            "--height", "720"
        ]
    }

    args(arguments)
    jvmArgs(jvmArguments)
}
```

Keep in mind that `build.gradle` is sometimes automatically updated and rewritten, so it may disappear or maybe they added a way to add cli arguments to `build.properties`  in the future 🤷‍♂️

*Mainly I just wanted to create this issue to help whoever else comes by later and give them a nice googleable bit of information that I lost a bunch of time to fixing it.*

### PR DESCRIPTION
**Pull request by eigenraven**
I think this is advanced enough to push for wider testing (but not merging yet!). The update to this buildscript version is more complex than usual, because:
 - Gradle version needs to be updated to 7.6
 - settings.gradle must be updated due to updated spotless configs (the old ones tried to reformat minecraft sources)
 - There is a chance some of the addon.gradle/dependencies.gradle scripts won't work anymore and need patching. I have added backwards compatibility for most of the constructs I could find, printing deprecation warnings as appropriate

What RFG brings compared to FG:
 - Better maintainability, it's been written from grounds up against the modern gradle apis, while reusing the pure java->java and similar transforms from FG for maximum compatibility by default
 - Support for running gradle with a newer jvm, while the code gets compiled to target version 8 via the toolchains feature
 - Optional [Jabel](https://github.com/bsideup/jabel) to support syntax of Java 17 targetting Java 8 bytecode
 - Updates LWJGL to 2.9.4 from 2.9.1 in the dev environment to fix some bugs
 - Adds `runObfClient`/`runObfServer` tasks to generate a launcher-like environment which runs obfuscated minecraft forge.
 - Optional generic injection, turning most of the visible raw types like `List`, `Map` in the decompiled code into nice `List<ItemStack>`, `Map<String, String>`. It's off by default for old code because it can cause compilation errors until code is updated.
 - Redone tag replacement. Replacement is supported for backwards compatibility, now done via a javac plugin to avoid 2 annoying bugs of FG: the old way didn't work in some IntelliJ configurations, leading to GRADLETOKEN_VERSION being visible or crashes; and the old method copied all sources to a separate directory which made following compiler errors annoying as they led you to the wrong copy of the file. A new method is also provided that generates a brand new class instead of modifying existing ones, it's less hacky and what most other Java projects do.
 - I understand RFG source code and can maintain it, better than trying to patch fixes into the ancient FG codebase

Update script (bash, linux, might work on macos) - to run, put it into the examplemod repo root and run it as `examplemod/update.sh which-mod-to-update/`:
```bash
#!/bin/bash

set -eou pipefail

SRC_FOLDER=$(readlink -f ${BASH_SOURCE%/*})
SRC_FOLDER=$(cd "${SRC_FOLDER}" && pwd -P)
TARGET_FOLDER=$(cd "$1" && pwd -P)

echo "Source at ${SRC_FOLDER}"
echo "Target at ${TARGET_FOLDER}"

if [[ "${SRC_FOLDER}" == "${TARGET_FOLDER}" ]]; then
  echo "Src=Target"
  exit 1
fi

if [[ ! -f "${TARGET_FOLDER}"/gradlew ]]; then
  echo "Target at ${TARGET_FOLDER} is not a gradle project"
  exit 1
fi

if [[ ! -f "${TARGET_FOLDER}"/gradlew ]]; then
  echo "Target at ${TARGET_FOLDER} is not a gradle project"
  exit 1
fi

if ! grep -Fq "GTNewHorizons/ExampleMod1.7.10/master/build.gradle" "${TARGET_FOLDER}"/build.gradle; then
  echo "Target is not using the ExampleMod buildscript"
  exit 1
fi

if grep -Fq "com.gtnewhorizons.retrofuturagradle" "${TARGET_FOLDER}"/build.gradle; then
  echo "Target is already using the RFG buildscript"
  exit 0 # No-op
fi

cd "${TARGET_FOLDER}/"
git checkout master
git pull
git reset --hard
git clean -f

cp -av "${SRC_FOLDER}/gradle" "${SRC_FOLDER}/gradlew" "${SRC_FOLDER}/gradlew.bat" "${SRC_FOLDER}/build.gradle" "${TARGET_FOLDER}/"
cp -av "${SRC_FOLDER}/settings.gradle.example" "${TARGET_FOLDER}/settings.gradle"
cp -av "${SRC_FOLDER}/.gitattributes" "${TARGET_FOLDER}/.gitattributes"
git add --chmod=+x ./gradlew
git commit -am "[ci skip] Update buildscript to RetroFuturaGradle"

./gradlew spotlessApply
git commit -am "[ci skip] spotlessApply with the new settings"

echo "# See https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view" > ".git-blame-ignore-revs"
echo "# Ignore spotlessApply reformat" > ".git-blame-ignore-revs"
git rev-parse HEAD >> ".git-blame-ignore-revs"
git add ".git-blame-ignore-revs"
git commit -am "[ci skip] Add git-blame-ignore-revs for spotlessApply"

git push origin master
```

Closes #64 
Closes #88

### CODE REVIEW

**Comment by ah-OOG-ah** [comment]
What is this?
image

**Comment by mitchej123** [comment]
> What is this? I'm getting it when I run runClient, and probably a few other times too. image

This is from CCC or CCL, and it's looking for the mappings directory so it can do some magic in dev

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -9,5 +9,5 @@ on:
 
 jobs:
   build-and-test:
-    uses: GTNewHorizons/GTNH-Actions-Workflows/.github/workflows/build-and-test.yml@master
+    uses: GTNewHorizons/GTNH-Actions-Workflows/.github/workflows/build-and-test.yml@rfg
```
Comment (eigenraven) [inline .github/workflows/build-and-test.yml:12]: The actions changes are temporary until we move mods over to the new buildscript, they're only needed for Jabel which is opt-in (just on by default for new projects including this template)

**Review by mitchej123 — APPROVED** [review]
:shipit: and see what happens

**Comment by mitchej123** [comment]
Note: needed to update the ExampleMod buildscript identification to `if ! grep -Eq "GTNewHorizons/ExampleMod1.7.10/(main|master)/build.gradle" ./build.gradle; then`

**Pull request merged**